{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6f577a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "225aa252",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS11</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1213983</td>\n",
       "      <td>1214153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS11</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001832731</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1213983</td>\n",
       "      <td>1214127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS11</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000368538</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1214125</td>\n",
       "      <td>1214127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS11</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature    start      end  score strand frame  \\\n",
       "0       1  ensembl_havana         gene  1211340  1214153    NaN      -     0   \n",
       "1       1  ensembl_havana   transcript  1211340  1214153    NaN      -     0   \n",
       "2       1  ensembl_havana         exon  1213983  1214153    NaN      -     0   \n",
       "3       1  ensembl_havana          CDS  1213983  1214127    NaN      -     0   \n",
       "4       1  ensembl_havana  start_codon  1214125  1214127    NaN      -     0   \n",
       "\n",
       "           gene_id gene_version  ... transcript_source transcript_biotype  \\\n",
       "0  ENSG00000186827           11  ...                                        \n",
       "1  ENSG00000186827           11  ...    ensembl_havana     protein_coding   \n",
       "2  ENSG00000186827           11  ...    ensembl_havana     protein_coding   \n",
       "3  ENSG00000186827           11  ...    ensembl_havana     protein_coding   \n",
       "4  ENSG00000186827           11  ...    ensembl_havana     protein_coding   \n",
       "\n",
       "          tag ccds_id transcript_support_level exon_number          exon_id  \\\n",
       "0                                                                             \n",
       "1  CCDS,basic  CCDS11                        1                                \n",
       "2  CCDS,basic  CCDS11                        1           1  ENSE00001832731   \n",
       "3  CCDS,basic  CCDS11                        1           1                    \n",
       "4  CCDS,basic  CCDS11                        1           1                    \n",
       "\n",
       "  exon_version       protein_id protein_version  \n",
       "0                                                \n",
       "1                                                \n",
       "2            2                                   \n",
       "3               ENSP00000368538               3  \n",
       "4                                                \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e57733",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "\n",
    "s = StringIO()\n",
    "x = [['1', 'havana', 'gene', '365389', '522928', '.', '-', '.', 'gene_id \"ENSG00000237094\"; gene_version \"12\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";'], ['1', 'havana', 'transcript', '501588', '517252', '.', '-', '.', 'gene_id \"ENSG00000237094\"; gene_version \"12\"; transcript_id \"ENST00000641303\"; transcript_version \"1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\"; transcript_source \"havana\"; transcript_biotype \"processed_transcript\";'], ['1', 'havana', 'transcript', '501604', '517225', '.', '-', '.', 'gene_id \"ENSG00000237094\"; gene_version \"12\"; transcript_id \"ENST00000641063\"; transcript_version \"1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\"; transcript_source \"havana\"; transcript_biotype \"processed_transcript\";'], ['1', 'havana', 'transcript', '504865', '522928', '.', '-', '.', 'gene_id \"ENSG00000237094\"; gene_version \"12\"; transcript_id \"ENST00000642124\"; transcript_version \"1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\"; transcript_source \"havana\"; transcript_biotype \"processed_transcript\"; tag \"basic\";'], ['1', 'havana', 'exon', '517156', '517225', '.', '-', '.', 'gene_id \"ENSG00000237094\"; gene_version \"12\"; transcript_id \"ENST00000641063\"; transcript_version \"1\"; exon_number \"1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\"; transcript_source \"havana\"; transcript_biotype \"processed_transcript\"; exon_id \"ENSE00003812919\"; exon_version \"1\";']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "f2f97bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "ucsc = pd.read_csv(\"grch38.knownGene.txt.gz\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a16226bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/a233287/pve/lib/python3.8/site-packages/gtfparse/read_gtf.py:151: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  result = parse_gtf(\n",
      "/Users/a233287/pve/lib/python3.8/site-packages/gtfparse/read_gtf.py:151: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  result = parse_gtf(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version']\n"
     ]
    }
   ],
   "source": [
    "ensmbl = read_gtf(\"Homo_sapiens.GRCh38.105.chr.gtf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a254a863",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name_table = pd.read_csv(\"gene_name_table.txt\", sep=\"\\t\")\n",
    "medgenes = pd.read_csv(\"MRG_medgenes_names.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "eda20795",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10111"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(medgenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9b94999c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4981 Redundant genes\n"
     ]
    }
   ],
   "source": [
    "x = medgenes[\"gene\"].value_counts()\n",
    "redund = x[x > 1]\n",
    "print(len(redund), \"Redundant genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "30c6ddec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5130"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes = medgenes.sort_values(['gene']).drop_duplicates()\n",
    "len(medgenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "71284e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "medgenes[\"has_table_query\"] = medgenes[\"gene\"].isin(gene_name_table[\"query\"]).astype(bool)\n",
    "medgenes[\"has_table_hit\"] = medgenes[\"gene\"].isin(gene_name_table[~gene_name_table[\"entrezgene\"].isna()][\"query\"]).astype(bool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f34fcd13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>has_table_hit</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_table_query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>44.0</td>\n",
       "      <td>5069.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "has_table_hit    False    True\n",
       "has_table_query               \n",
       "False             17.0     NaN\n",
       "True              44.0  5069.0"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes.groupby([\"has_table_query\", \"has_table_hit\"]).size().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "f0179fc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "medgenes.loc[medgenes[\"has_ensmbl_hit\"], 'ensmbl_name'] = medgenes.loc[medgenes[\"has_ensmbl_hit\"]]['gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "11d64e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "view = ~medgenes[\"has_ensmbl_hit\"] & medgenes['symbol_has_ensmbl_hit']\n",
    "medgenes.loc[view, 'ensmbl_name'] = medgenes.loc[view]['symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "990a0937",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>has_table_query</th>\n",
       "      <th>has_table_hit</th>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "      <th>symbol</th>\n",
       "      <th>symbol_has_ensmbl_hit</th>\n",
       "      <th>has_ucsc_geneName</th>\n",
       "      <th>symbol_has_ucsc_geneName</th>\n",
       "      <th>ensmbl_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2521</th>\n",
       "      <td>C10orf11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579</th>\n",
       "      <td>C10orf2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8261</th>\n",
       "      <td>C12orf65</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>C2orf71</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>C4orf26</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351</th>\n",
       "      <td>C5orf42</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7241</th>\n",
       "      <td>C8orf37</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8986</th>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3939</th>\n",
       "      <td>CCL4L1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CCL4L1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4665</th>\n",
       "      <td>GSTT1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028</th>\n",
       "      <td>PRSS3P2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>PRSS3P2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1800</th>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8049</th>\n",
       "      <td>TAS2R45</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>TAS2R45</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4150</th>\n",
       "      <td>TCEB3C</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ELOA3BP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene  has_table_query  has_table_hit  has_ensmbl_hit   symbol  \\\n",
       "2521  C10orf11            False          False           False      NaN   \n",
       "2579   C10orf2            False          False           False      NaN   \n",
       "8261  C12orf65            False          False           False      NaN   \n",
       "493    C2orf71            False          False           False      NaN   \n",
       "1205   C4orf26            False          False           False      NaN   \n",
       "1351   C5orf42            False          False           False      NaN   \n",
       "7241   C8orf37            False          False           False      NaN   \n",
       "8986    CCL3L1             True           True           False   CCL3L1   \n",
       "3939    CCL4L1             True           True           False   CCL4L1   \n",
       "4665     GSTT1             True           True           False    GSTT1   \n",
       "2028   PRSS3P2             True           True           False  PRSS3P2   \n",
       "1800         T            False          False           False      NaN   \n",
       "8049   TAS2R45             True           True           False  TAS2R45   \n",
       "4150    TCEB3C             True           True           False  ELOA3BP   \n",
       "\n",
       "     symbol_has_ensmbl_hit  has_ucsc_geneName  symbol_has_ucsc_geneName  \\\n",
       "2521                   NaN              False                     False   \n",
       "2579                   NaN              False                     False   \n",
       "8261                   NaN               True                     False   \n",
       "493                    NaN              False                     False   \n",
       "1205                   NaN              False                     False   \n",
       "1351                   NaN              False                     False   \n",
       "7241                   NaN               True                     False   \n",
       "8986                 False               True                      True   \n",
       "3939                 False              False                     False   \n",
       "4665                 False               True                      True   \n",
       "2028                 False               True                      True   \n",
       "1800                   NaN              False                     False   \n",
       "8049                 False               True                      True   \n",
       "4150                 False              False                     False   \n",
       "\n",
       "     ensmbl_name  \n",
       "2521         NaN  \n",
       "2579         NaN  \n",
       "8261         NaN  \n",
       "493          NaN  \n",
       "1205         NaN  \n",
       "1351         NaN  \n",
       "7241         NaN  \n",
       "8986         NaN  \n",
       "3939         NaN  \n",
       "4665         NaN  \n",
       "2028         NaN  \n",
       "1800         NaN  \n",
       "8049         NaN  \n",
       "4150         NaN  "
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing = medgenes[medgenes['ensmbl_name'].isna()]\n",
    "missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "f8d76d00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "54fc88e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#medgene - ensmbl_id - ensmbl_name - Notes\n",
    "# C10orf11 - ENSG00000148655 - LRMDA\n",
    "# C10orf2 - ENSG00000107815 - TWNK\n",
    "# C12orf65 - ENSG00000130921 - MTRFR\n",
    "# C2orf71 - ENSG00000179270 - PCARE\n",
    "# C4orf26 - ENSG00000174792 - ODAPH\n",
    "# C5orf42 - ENSG00000197603 - CPLANE1\n",
    "# C8orf37 - ENSG00000156172 - CFAP418\n",
    "# CCL3L1 -  ENSG00000276085 - CCL3L3\n",
    "# CCL4L1 - Missing - - Might be ENSG00000276070, CCL4L2\n",
    "# GSTT1 - Missing - - Closest ENSG00000277656\n",
    "# PRSS3P2 - Missing - - Closest ENSG00000282667\n",
    "# T - Missing - - No clue.. ENSG00000166349 / RAG1??\n",
    "# TAS2R45 - Missing - - Closest ENSG00000261936\n",
    "# TCEB3C - Missing - - Closest ENSG00000275553 / ELOA3CP\n",
    "manual_translate = {\"C10orf11\" :\"LRMDA\",\n",
    "\"C10orf2\" :\"TWNK\",\n",
    "\"C12orf65\" :\"MTRFR\",\n",
    "\"C2orf71\" :\"PCARE\",\n",
    "\"C4orf26\" :\"ODAPH\",\n",
    "\"C5orf42\" :\"CPLANE1\",\n",
    "\"C8orf37\" :\"CFAP418\",\n",
    "\"CCL3L1\" :\"CCL3L3\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "fa62d01a",
   "metadata": {},
   "outputs": [],
   "source": [
    "medgenes.loc[medgenes['ensmbl_name'].isna(), 'ensmbl_name'] = medgenes.loc[medgenes['ensmbl_name'].isna(), 'gene'].map(manual_translate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "85138f22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>has_table_query</th>\n",
       "      <th>has_table_hit</th>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "      <th>symbol</th>\n",
       "      <th>symbol_has_ensmbl_hit</th>\n",
       "      <th>has_ucsc_geneName</th>\n",
       "      <th>symbol_has_ucsc_geneName</th>\n",
       "      <th>ensmbl_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3939</th>\n",
       "      <td>CCL4L1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CCL4L1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4665</th>\n",
       "      <td>GSTT1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028</th>\n",
       "      <td>PRSS3P2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>PRSS3P2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1800</th>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8049</th>\n",
       "      <td>TAS2R45</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>TAS2R45</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4150</th>\n",
       "      <td>TCEB3C</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ELOA3BP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  has_table_query  has_table_hit  has_ensmbl_hit   symbol  \\\n",
       "3939   CCL4L1             True           True           False   CCL4L1   \n",
       "4665    GSTT1             True           True           False    GSTT1   \n",
       "2028  PRSS3P2             True           True           False  PRSS3P2   \n",
       "1800        T            False          False           False      NaN   \n",
       "8049  TAS2R45             True           True           False  TAS2R45   \n",
       "4150   TCEB3C             True           True           False  ELOA3BP   \n",
       "\n",
       "     symbol_has_ensmbl_hit  has_ucsc_geneName  symbol_has_ucsc_geneName  \\\n",
       "3939                 False              False                     False   \n",
       "4665                 False               True                      True   \n",
       "2028                 False               True                      True   \n",
       "1800                   NaN              False                     False   \n",
       "8049                 False               True                      True   \n",
       "4150                 False              False                     False   \n",
       "\n",
       "     ensmbl_name  \n",
       "3939         NaN  \n",
       "4665         NaN  \n",
       "2028         NaN  \n",
       "1800         NaN  \n",
       "8049         NaN  \n",
       "4150         NaN  "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing = medgenes[medgenes['ensmbl_name'].isna()]\n",
    "missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "88122d4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5130, 5124)"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(medgenes), len(medgenes[~medgenes[\"ensmbl_name\"].isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "d9ded827",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(medgenes[\"ensmbl_name\"].unique(), name=\"ensmbl_name\").to_csv(\"medgene_ensmbl_names.txt\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "ca4b5fb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7853      A1CF\n",
       "2992       A2M\n",
       "4732    A4GALT\n",
       "1016     A4GNT\n",
       "8135      AAAS\n",
       "Name: ensmbl_name, dtype: object"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes[\"ensmbl_name\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "60e93fa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon',\n",
       "       'five_prime_utr', 'three_prime_utr', 'Selenocysteine'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl['feature'].unique()\n",
    "# | -------  gene  -------- |\n",
    "#    | --- transcript --- |\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "2d9cafbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>581521</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>138123713</td>\n",
       "      <td>138132390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname          source feature      start        end  score strand  \\\n",
       "581521       3  ensembl_havana    gene  138123713  138132390    NaN      -   \n",
       "\n",
       "       frame          gene_id gene_version  ... transcript_source  \\\n",
       "581521     0  ENSG00000118017            4  ...                     \n",
       "\n",
       "       transcript_biotype tag ccds_id transcript_support_level exon_number  \\\n",
       "581521                                                                       \n",
       "\n",
       "       exon_id exon_version protein_id protein_version  \n",
       "581521                                                  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl[(ensmbl[\"gene_name\"] == \"A4GNT\") & (ensmbl['feature'] == 'gene')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "571e75ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>581522</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>138123713</td>\n",
       "      <td>138132390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS3097</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname          source     feature      start        end  score  \\\n",
       "581522       3  ensembl_havana  transcript  138123713  138132390    NaN   \n",
       "\n",
       "       strand frame          gene_id gene_version  ... transcript_source  \\\n",
       "581522      -     0  ENSG00000118017            4  ...    ensembl_havana   \n",
       "\n",
       "       transcript_biotype         tag   ccds_id transcript_support_level  \\\n",
       "581522     protein_coding  CCDS,basic  CCDS3097                        1   \n",
       "\n",
       "       exon_number exon_id exon_version protein_id protein_version  \n",
       "581522                                                              \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl[(ensmbl[\"gene_name\"] == \"A4GNT\") & (ensmbl['feature'] == 'transcript')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "afcc46d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>581523</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>138132212</td>\n",
       "      <td>138132390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS3097</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001241635</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581524</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>138130849</td>\n",
       "      <td>138131282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS3097</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001241655</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581527</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>138123713</td>\n",
       "      <td>138124878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>CCDS3097</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00001037446</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname          source feature      start        end  score strand  \\\n",
       "581523       3  ensembl_havana    exon  138132212  138132390    NaN      -   \n",
       "581524       3  ensembl_havana    exon  138130849  138131282    NaN      -   \n",
       "581527       3  ensembl_havana    exon  138123713  138124878    NaN      -   \n",
       "\n",
       "       frame          gene_id gene_version  ... transcript_source  \\\n",
       "581523     0  ENSG00000118017            4  ...    ensembl_havana   \n",
       "581524     0  ENSG00000118017            4  ...    ensembl_havana   \n",
       "581527     0  ENSG00000118017            4  ...    ensembl_havana   \n",
       "\n",
       "       transcript_biotype         tag   ccds_id transcript_support_level  \\\n",
       "581523     protein_coding  CCDS,basic  CCDS3097                        1   \n",
       "581524     protein_coding  CCDS,basic  CCDS3097                        1   \n",
       "581527     protein_coding  CCDS,basic  CCDS3097                        1   \n",
       "\n",
       "       exon_number          exon_id exon_version protein_id protein_version  \n",
       "581523           1  ENSE00001241635            3                             \n",
       "581524           2  ENSE00001241655            1                             \n",
       "581527           3  ENSE00001037446            6                             \n",
       "\n",
       "[3 rows x 26 columns]"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl[(ensmbl[\"gene_name\"] == \"A4GNT\") & (ensmbl['feature'] == 'exon')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "70622e2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
       "       'frame', 'gene_id', 'gene_version', 'gene_name', 'gene_source',\n",
       "       'gene_biotype', 'transcript_id', 'transcript_version',\n",
       "       'transcript_name', 'transcript_source', 'transcript_biotype', 'tag',\n",
       "       'ccds_id', 'transcript_support_level', 'exon_number', 'exon_id',\n",
       "       'exon_version', 'protein_id', 'protein_version'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "1c9a1bc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature</th>\n",
       "      <th>CDS</th>\n",
       "      <th>Selenocysteine</th>\n",
       "      <th>exon</th>\n",
       "      <th>five_prime_utr</th>\n",
       "      <th>gene</th>\n",
       "      <th>start_codon</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>three_prime_utr</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>81.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>47.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>59.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>240.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF81</th>\n",
       "      <td>14.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF3</th>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP</th>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP2</th>\n",
       "      <td>27.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRSR2</th>\n",
       "      <td>50.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5110 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "feature     CDS  Selenocysteine   exon  five_prime_utr  gene  start_codon  \\\n",
       "gene_name                                                                   \n",
       "A1CF       81.0             NaN  111.0            26.0   1.0          8.0   \n",
       "A2M        47.0             NaN   87.0             4.0   1.0          3.0   \n",
       "A4GALT      4.0             NaN   17.0            10.0   1.0          4.0   \n",
       "A4GNT       2.0             NaN    3.0             2.0   1.0          1.0   \n",
       "AAAS       59.0             NaN  240.0            13.0   1.0          4.0   \n",
       "...         ...             ...    ...             ...   ...          ...   \n",
       "ZNF81      14.0             NaN   22.0            10.0   1.0          4.0   \n",
       "ZNRF3      25.0             NaN   26.0             3.0   1.0          3.0   \n",
       "ZPBP       20.0             NaN   31.0             6.0   1.0          4.0   \n",
       "ZPBP2      27.0             NaN   27.0             4.0   1.0          4.0   \n",
       "ZRSR2      50.0             NaN   56.0             5.0   1.0          5.0   \n",
       "\n",
       "feature    stop_codon  three_prime_utr  transcript  \n",
       "gene_name                                           \n",
       "A1CF              7.0              7.0        10.0  \n",
       "A2M               3.0              9.0        13.0  \n",
       "A4GALT            4.0              4.0         7.0  \n",
       "A4GNT             1.0              1.0         1.0  \n",
       "AAAS              4.0              5.0        22.0  \n",
       "...               ...              ...         ...  \n",
       "ZNF81             3.0              3.0         5.0  \n",
       "ZNRF3             3.0              3.0         3.0  \n",
       "ZPBP              3.0              4.0         6.0  \n",
       "ZPBP2             3.0              3.0         4.0  \n",
       "ZRSR2             5.0              7.0         6.0  \n",
       "\n",
       "[5110 rows x 9 columns]"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensmbl[ensmbl['gene_name'].isin(medgenes['ensmbl_name'])].groupby(['gene_name', 'feature']).size().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "0bf490be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>has_table_query</th>\n",
       "      <th>has_table_hit</th>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "      <th>symbol</th>\n",
       "      <th>symbol_has_ensmbl_hit</th>\n",
       "      <th>has_ucsc_geneName</th>\n",
       "      <th>symbol_has_ucsc_geneName</th>\n",
       "      <th>ensmbl_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2992</th>\n",
       "      <td>A2M</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A2M</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  has_table_query  has_table_hit  has_ensmbl_hit symbol  \\\n",
       "2992  A2M             True           True            True    A2M   \n",
       "\n",
       "     symbol_has_ensmbl_hit  has_ucsc_geneName  symbol_has_ucsc_geneName  \\\n",
       "2992                  True               True                      True   \n",
       "\n",
       "     ensmbl_name  \n",
       "2992         A2M  "
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes[medgenes[\"ensmbl_name\"] == 'A2M']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "daf20cdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     104\n",
       "False      6\n",
       "Name: symbol_has_ensmbl_hit, dtype: int64"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes[~medgenes[\"has_ensmbl_hit\"]]['symbol_has_ensmbl_hit'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "3073b62f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     5012\n",
       "False     118\n",
       "Name: has_ensmbl_hit, dtype: int64"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes[\"has_ensmbl_hit\"] = medgenes['gene'].isin(ensmbl[\"gene_name\"])\n",
    "medgenes['has_ensmbl_hit'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "1318bc72",
   "metadata": {},
   "outputs": [],
   "source": [
    "medgenes[\"has_ensmbl_hit\"] = medgenes[medgenes[\"has_table_hit\"]]['gene'].isin(ensmbl[\"gene_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2794057",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensmbl_gene_name = medgenes[medgenes[\"has_table_query\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "1f27f415",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>has_table_query</th>\n",
       "      <th>has_table_hit</th>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3779</th>\n",
       "      <td>AARS</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>ADCK3</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4672</th>\n",
       "      <td>ADRBK2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2349</th>\n",
       "      <td>AKAP2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4758</th>\n",
       "      <td>ARSE</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4646</th>\n",
       "      <td>UFD1L</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1894</th>\n",
       "      <td>WBSCR17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1118</th>\n",
       "      <td>WHSC1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1729</th>\n",
       "      <td>WISP3</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>YARS</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  has_table_query  has_table_hit has_ensmbl_hit\n",
       "3779     AARS             True           True          False\n",
       "422     ADCK3             True           True          False\n",
       "4672   ADRBK2             True           True          False\n",
       "2349    AKAP2             True           True          False\n",
       "4758     ARSE             True           True          False\n",
       "...       ...              ...            ...            ...\n",
       "4646    UFD1L             True           True          False\n",
       "1894  WBSCR17             True           True          False\n",
       "1118    WHSC1             True           True          False\n",
       "1729    WISP3             True           True          False\n",
       "93       YARS             True           True          False\n",
       "\n",
       "[110 rows x 4 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes[(medgenes['has_ensmbl_hit'] == False) & (medgenes['has_table_hit'] == True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "0a44dd85",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name_table['alias'] = gene_name_table['alias'].astype(str).apply(lambda x: [_ for _ in x.split(', ') if _ != 'nan'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "3016cd1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = gene_name_table.set_index('query')\n",
    "medgenes['symbol'] = medgenes['gene'].map(tmp['symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "1548ef6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9732943469785575"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(medgenes['gene'] == medgenes['symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "23cd709b",
   "metadata": {},
   "outputs": [],
   "source": [
    "medgenes[\"symbol_has_ensmbl_hit\"] = medgenes[medgenes[\"has_table_hit\"]]['symbol'].isin(ensmbl[\"gene_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "79cbb604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>symbol_has_ensmbl_hit</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>6</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>1</td>\n",
       "      <td>4958</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "symbol_has_ensmbl_hit  False  True\n",
       "has_ensmbl_hit                    \n",
       "False                      6   104\n",
       "True                       1  4958"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes.groupby(['has_ensmbl_hit', 'symbol_has_ensmbl_hit']).size().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "c246a189",
   "metadata": {},
   "outputs": [],
   "source": [
    "# has ucsc\n",
    "medgenes['has_ucsc_geneName'] = medgenes['gene'].isin(ucsc[\"geneName\"])\n",
    "medgenes['symbol_has_ucsc_geneName'] = medgenes['symbol'].isin(ucsc['geneName'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "56e815a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>has_table_query</th>\n",
       "      <th>has_table_hit</th>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "      <th>symbol</th>\n",
       "      <th>symbol_has_ensmbl_hit</th>\n",
       "      <th>has_ucsc_geneName</th>\n",
       "      <th>symbol_has_ucsc_geneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3939</th>\n",
       "      <td>CCL4L1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CCL4L1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4150</th>\n",
       "      <td>TCEB3C</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ELOA3BP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  has_table_query  has_table_hit has_ensmbl_hit   symbol  \\\n",
       "3939  CCL4L1             True           True          False   CCL4L1   \n",
       "4150  TCEB3C             True           True          False  ELOA3BP   \n",
       "\n",
       "     symbol_has_ensmbl_hit  has_ucsc_geneName  symbol_has_ucsc_geneName  \n",
       "3939                 False              False                     False  \n",
       "4150                 False              False                     False  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing = medgenes[(medgenes[\"has_ensmbl_hit\"] == False) & (medgenes['symbol_has_ensmbl_hit'] == False) &\n",
    "                  (medgenes[\"has_ucsc_geneName\"] == False) & (medgenes['symbol_has_ucsc_geneName'] == False)]\n",
    "missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "150f048d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCL4L1 False False\n",
      "AT744.2 False False\n",
      "CCL4L False False\n",
      "LAG-1 False False\n",
      "LAG1 False False\n",
      "MIP-1-beta False False\n",
      "\n",
      "TCEB3C False False\n",
      "ELOA3 False False\n",
      "ELOA3B False False\n",
      "ELOA3L1 False False\n",
      "EloA3C False False\n",
      "Elongin-A3 False False\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for _ in [\"CCL4L1\", \"TCEB3C\"]:\n",
    "    print(_, _ in ucsc['geneName'], _ in ensmbl[\"gene_name\"])\n",
    "    for alias in gene_name_table[gene_name_table[\"query\"] == _]['alias'].iloc[0]:\n",
    "        print(alias, alias in ucsc['geneName'], alias in ensmbl[\"gene_name\"])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "d5dcffc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensmbl_manual_lookup = {\"CCL4L1\": \"ENSG00000276070\",\n",
    "                        \"TCEB3C\":\"ENSG00000288607\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c95de64f",
   "metadata": {},
   "source": [
    "okay - now I need to go through the 5 pieces of keys and try to pull them each out.\n",
    "Actually, wait - I need to take the *has_ucsc_geneName and try to map them to Ensmbl so I have a unified coordinate system"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "d288d87a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#chrom', 'chromStart', 'chromEnd', 'name', 'score', 'strand',\n",
       "       'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes',\n",
       "       'chromStarts', 'name2', 'cdsStartStat', 'cdsEndStat', 'exonFrames',\n",
       "       'type', 'geneName', 'geneName2', 'geneType', 'transcriptClass',\n",
       "       'source', 'transcriptType', 'tag', 'level', 'tier'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ucsc.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "db27cc26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>reserved</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>...</th>\n",
       "      <th>type</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneName2</th>\n",
       "      <th>geneType</th>\n",
       "      <th>transcriptClass</th>\n",
       "      <th>source</th>\n",
       "      <th>transcriptType</th>\n",
       "      <th>tag</th>\n",
       "      <th>level</th>\n",
       "      <th>tier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>196412</th>\n",
       "      <td>chr17</td>\n",
       "      <td>36194868</td>\n",
       "      <td>36196747</td>\n",
       "      <td>ENST00000612839.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>36194868</td>\n",
       "      <td>36194868</td>\n",
       "      <td>16646144</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>retained_intron_CDS</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196413</th>\n",
       "      <td>chr17</td>\n",
       "      <td>36194868</td>\n",
       "      <td>36196758</td>\n",
       "      <td>ENST00000619989.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>36195285</td>\n",
       "      <td>36196673</td>\n",
       "      <td>789624</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>P16619</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>ensembl_havana_transcript_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,Ensembl_canonical,MANE_Select,appris_prin...</td>\n",
       "      <td>2</td>\n",
       "      <td>canonical,basic,all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196414</th>\n",
       "      <td>chr17</td>\n",
       "      <td>36194995</td>\n",
       "      <td>36196457</td>\n",
       "      <td>ENST00000616702.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>36194995</td>\n",
       "      <td>36194995</td>\n",
       "      <td>25600</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248196</th>\n",
       "      <td>chr7_KI270803v1_alt</td>\n",
       "      <td>773585</td>\n",
       "      <td>777110</td>\n",
       "      <td>ENST00000634192.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>773585</td>\n",
       "      <td>773585</td>\n",
       "      <td>16724991</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>PRSS3P2</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>Ensembl_canonical,basic</td>\n",
       "      <td>2</td>\n",
       "      <td>canonical,basic,all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250652</th>\n",
       "      <td>chr12_GL877876v1_alt</td>\n",
       "      <td>327494</td>\n",
       "      <td>328424</td>\n",
       "      <td>ENST00000571573.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>327494</td>\n",
       "      <td>327494</td>\n",
       "      <td>16724991</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>TAS2R45</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>Ensembl_canonical,basic,overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>canonical,basic,all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257953</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270313</td>\n",
       "      <td>278434</td>\n",
       "      <td>ENST00000628171.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>271169</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JQM0</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257954</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270313</td>\n",
       "      <td>278442</td>\n",
       "      <td>ENST00000621513.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273663</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JMS2</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257955</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270313</td>\n",
       "      <td>278459</td>\n",
       "      <td>ENST00000612885.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>270597</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>P30711</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>ensembl_havana_transcript_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Ensembl_canonical,basic</td>\n",
       "      <td>2</td>\n",
       "      <td>canonical,basic,all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257956</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270498</td>\n",
       "      <td>278440</td>\n",
       "      <td>ENST00000629554.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>271147</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JQD8</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257957</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270567</td>\n",
       "      <td>278421</td>\n",
       "      <td>ENST00000628969.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>270597</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JRQ5</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>2</td>\n",
       "      <td>basic,all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257958</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270661</td>\n",
       "      <td>278423</td>\n",
       "      <td>ENST00000628470.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>270674</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JRN4</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>2</td>\n",
       "      <td>basic,all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257959</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270707</td>\n",
       "      <td>278427</td>\n",
       "      <td>ENST00000627958.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>270707</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A1B0GXB6</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>cds_end_NF,mRNA_end_NF</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257960</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>270707</td>\n",
       "      <td>278427</td>\n",
       "      <td>ENST00000630428.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273596</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JRG0</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257961</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>271089</td>\n",
       "      <td>278446</td>\n",
       "      <td>ENST00000628465.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>273650</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JRJ5</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257962</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>274048</td>\n",
       "      <td>278486</td>\n",
       "      <td>ENST00000631014.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>274048</td>\n",
       "      <td>274048</td>\n",
       "      <td>16646144</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257963</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>274059</td>\n",
       "      <td>278436</td>\n",
       "      <td>ENST00000630911.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>274059</td>\n",
       "      <td>274059</td>\n",
       "      <td>16646144</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257964</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>275879</td>\n",
       "      <td>278855</td>\n",
       "      <td>ENST00000628417.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>275879</td>\n",
       "      <td>278406</td>\n",
       "      <td>789624</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>A0A0G2JQD2</td>\n",
       "      <td>none</td>\n",
       "      <td>coding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>alternative_5_UTR,cds_end_NF,mRNA_end_NF</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257965</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>276444</td>\n",
       "      <td>278446</td>\n",
       "      <td>ENST00000626730.2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>276444</td>\n",
       "      <td>276444</td>\n",
       "      <td>16646144</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257966</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>277107</td>\n",
       "      <td>278446</td>\n",
       "      <td>ENST00000628175.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>277107</td>\n",
       "      <td>277107</td>\n",
       "      <td>16646144</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257967</th>\n",
       "      <td>chr22_KI270879v1_alt</td>\n",
       "      <td>277294</td>\n",
       "      <td>278445</td>\n",
       "      <td>ENST00000625361.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>277294</td>\n",
       "      <td>277294</td>\n",
       "      <td>16646144</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>GSTT1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>nonCoding</td>\n",
       "      <td>havana_homo_sapiens</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>none</td>\n",
       "      <td>2</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      #chrom  chromStart  chromEnd               name  score  \\\n",
       "196412                 chr17    36194868  36196747  ENST00000612839.1      0   \n",
       "196413                 chr17    36194868  36196758  ENST00000619989.1      0   \n",
       "196414                 chr17    36194995  36196457  ENST00000616702.1      0   \n",
       "248196   chr7_KI270803v1_alt      773585    777110  ENST00000634192.1      0   \n",
       "250652  chr12_GL877876v1_alt      327494    328424  ENST00000571573.1      0   \n",
       "257953  chr22_KI270879v1_alt      270313    278434  ENST00000628171.2      0   \n",
       "257954  chr22_KI270879v1_alt      270313    278442  ENST00000621513.3      0   \n",
       "257955  chr22_KI270879v1_alt      270313    278459  ENST00000612885.3      0   \n",
       "257956  chr22_KI270879v1_alt      270498    278440  ENST00000629554.1      0   \n",
       "257957  chr22_KI270879v1_alt      270567    278421  ENST00000628969.2      0   \n",
       "257958  chr22_KI270879v1_alt      270661    278423  ENST00000628470.2      0   \n",
       "257959  chr22_KI270879v1_alt      270707    278427  ENST00000627958.1      0   \n",
       "257960  chr22_KI270879v1_alt      270707    278427  ENST00000630428.2      0   \n",
       "257961  chr22_KI270879v1_alt      271089    278446  ENST00000628465.1      0   \n",
       "257962  chr22_KI270879v1_alt      274048    278486  ENST00000631014.1      0   \n",
       "257963  chr22_KI270879v1_alt      274059    278436  ENST00000630911.1      0   \n",
       "257964  chr22_KI270879v1_alt      275879    278855  ENST00000628417.1      0   \n",
       "257965  chr22_KI270879v1_alt      276444    278446  ENST00000626730.2      0   \n",
       "257966  chr22_KI270879v1_alt      277107    278446  ENST00000628175.1      0   \n",
       "257967  chr22_KI270879v1_alt      277294    278445  ENST00000625361.1      0   \n",
       "\n",
       "       strand  thickStart  thickEnd  reserved  blockCount  ...  type geneName  \\\n",
       "196412      -    36194868  36194868  16646144           2  ...  none   CCL3L1   \n",
       "196413      -    36195285  36196673    789624           3  ...  none   CCL3L1   \n",
       "196414      -    36194995  36194995     25600           3  ...  none   CCL3L1   \n",
       "248196      +      773585    773585  16724991           5  ...  none  PRSS3P2   \n",
       "250652      -      327494    327494  16724991           1  ...  none  TAS2R45   \n",
       "257953      -      271169    278406    789624           5  ...  none    GSTT1   \n",
       "257954      -      273663    278406    789624           4  ...  none    GSTT1   \n",
       "257955      -      270597    278406    789624           5  ...  none    GSTT1   \n",
       "257956      -      271147    278406    789624           3  ...  none    GSTT1   \n",
       "257957      -      270597    278406    789624           5  ...  none    GSTT1   \n",
       "257958      -      270674    278406    789624           3  ...  none    GSTT1   \n",
       "257959      -      270707    278406    789624           2  ...  none    GSTT1   \n",
       "257960      -      273596    278406    789624           4  ...  none    GSTT1   \n",
       "257961      -      273650    278406    789624           5  ...  none    GSTT1   \n",
       "257962      -      274048    274048  16646144           2  ...  none    GSTT1   \n",
       "257963      -      274059    274059  16646144           1  ...  none    GSTT1   \n",
       "257964      -      275879    278406    789624           3  ...  none    GSTT1   \n",
       "257965      -      276444    276444  16646144           2  ...  none    GSTT1   \n",
       "257966      -      277107    277107  16646144           2  ...  none    GSTT1   \n",
       "257967      -      277294    277294  16646144           2  ...  none    GSTT1   \n",
       "\n",
       "         geneName2 geneType transcriptClass  \\\n",
       "196412        none     none       nonCoding   \n",
       "196413      P16619     none          coding   \n",
       "196414        none     none       nonCoding   \n",
       "248196        none     none          pseudo   \n",
       "250652        none     none          pseudo   \n",
       "257953  A0A0G2JQM0     none          coding   \n",
       "257954  A0A0G2JMS2     none          coding   \n",
       "257955      P30711     none          coding   \n",
       "257956  A0A0G2JQD8     none          coding   \n",
       "257957  A0A0G2JRQ5     none          coding   \n",
       "257958  A0A0G2JRN4     none          coding   \n",
       "257959  A0A1B0GXB6     none          coding   \n",
       "257960  A0A0G2JRG0     none          coding   \n",
       "257961  A0A0G2JRJ5     none          coding   \n",
       "257962        none     none       nonCoding   \n",
       "257963        none     none       nonCoding   \n",
       "257964  A0A0G2JQD2     none          coding   \n",
       "257965        none     none       nonCoding   \n",
       "257966        none     none       nonCoding   \n",
       "257967        none     none       nonCoding   \n",
       "\n",
       "                                        source           transcriptType  \\\n",
       "196412                     havana_homo_sapiens          retained_intron   \n",
       "196413  ensembl_havana_transcript_homo_sapiens           protein_coding   \n",
       "196414                     havana_homo_sapiens     processed_transcript   \n",
       "248196                     havana_homo_sapiens   unprocessed_pseudogene   \n",
       "250652                     havana_homo_sapiens   unprocessed_pseudogene   \n",
       "257953                     havana_homo_sapiens  nonsense_mediated_decay   \n",
       "257954                     havana_homo_sapiens  nonsense_mediated_decay   \n",
       "257955  ensembl_havana_transcript_homo_sapiens           protein_coding   \n",
       "257956                     havana_homo_sapiens  nonsense_mediated_decay   \n",
       "257957                     havana_homo_sapiens           protein_coding   \n",
       "257958                     havana_homo_sapiens           protein_coding   \n",
       "257959                     havana_homo_sapiens           protein_coding   \n",
       "257960                     havana_homo_sapiens  nonsense_mediated_decay   \n",
       "257961                     havana_homo_sapiens  nonsense_mediated_decay   \n",
       "257962                     havana_homo_sapiens          retained_intron   \n",
       "257963                     havana_homo_sapiens          retained_intron   \n",
       "257964                     havana_homo_sapiens           protein_coding   \n",
       "257965                     havana_homo_sapiens          retained_intron   \n",
       "257966                     havana_homo_sapiens          retained_intron   \n",
       "257967                     havana_homo_sapiens          retained_intron   \n",
       "\n",
       "                                                      tag level  \\\n",
       "196412                                retained_intron_CDS     2   \n",
       "196413  CCDS,Ensembl_canonical,MANE_Select,appris_prin...     2   \n",
       "196414                                               none     2   \n",
       "248196                            Ensembl_canonical,basic     2   \n",
       "250652          Ensembl_canonical,basic,overlapping_locus     2   \n",
       "257953                                               none     2   \n",
       "257954                                               none     2   \n",
       "257955                            Ensembl_canonical,basic     2   \n",
       "257956                                               none     2   \n",
       "257957                                              basic     2   \n",
       "257958                                              basic     2   \n",
       "257959                             cds_end_NF,mRNA_end_NF     2   \n",
       "257960                                               none     2   \n",
       "257961                                               none     2   \n",
       "257962                                               none     2   \n",
       "257963                                               none     2   \n",
       "257964           alternative_5_UTR,cds_end_NF,mRNA_end_NF     2   \n",
       "257965                                               none     2   \n",
       "257966                                               none     2   \n",
       "257967                                               none     2   \n",
       "\n",
       "                       tier  \n",
       "196412                  all  \n",
       "196413  canonical,basic,all  \n",
       "196414                  all  \n",
       "248196  canonical,basic,all  \n",
       "250652  canonical,basic,all  \n",
       "257953                  all  \n",
       "257954                  all  \n",
       "257955  canonical,basic,all  \n",
       "257956                  all  \n",
       "257957            basic,all  \n",
       "257958            basic,all  \n",
       "257959                  all  \n",
       "257960                  all  \n",
       "257961                  all  \n",
       "257962                  all  \n",
       "257963                  all  \n",
       "257964                  all  \n",
       "257965                  all  \n",
       "257966                  all  \n",
       "257967                  all  \n",
       "\n",
       "[20 rows x 26 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "view = medgenes[(medgenes[\"has_ensmbl_hit\"] == False) & (medgenes['symbol_has_ensmbl_hit'] == False) &\n",
    "                   (medgenes[\"has_ucsc_geneName\"] == True)]\n",
    "ucsc[ucsc[\"geneName\"].isin(view[\"gene\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "45bf24b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>has_table_query</th>\n",
       "      <th>has_table_hit</th>\n",
       "      <th>has_ensmbl_hit</th>\n",
       "      <th>symbol</th>\n",
       "      <th>symbol_has_ensmbl_hit</th>\n",
       "      <th>has_ucsc_geneName</th>\n",
       "      <th>symbol_has_ucsc_geneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, has_table_query, has_table_hit, has_ensmbl_hit, symbol, symbol_has_ensmbl_hit, has_ucsc_geneName, symbol_has_ucsc_geneName]\n",
       "Index: []"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgenes[(medgenes[\"has_ensmbl_hit\"] == False) & (medgenes['symbol_has_ensmbl_hit'] == False) &\n",
    "                  (medgenes[\"has_ucsc_geneName\"] == False) & (medgenes['symbol_has_ucsc_geneName'] == True)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f15eba2d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
